src/Common/src/ThemingScope.cs

```diff
@@ -0,0 +1,92 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Runtime.InteropServices;
+using static Interop;
+
+namespace System.Windows.Forms
+{
+    /// <summary>
+    ///  This class provides static methods to create, activate and deactivate the theming scope.
+    /// </summary>
+    internal class ThemingScope
+    {
+        private static Kernel32.ACTCTXW s_enableThemingActivationContext;
+        private static IntPtr s_hActCtx;
+        private static bool s_contextCreationSucceeded;
+
+        /// <summary>
+        ///  We now use explicitactivate everywhere and use this method to determine if we
+        ///  really need to activate the activationcontext.  This should be pretty fast.
+        /// </summary>
+        private static bool IsContextActive()
+        {
+            if (s_contextCreationSucceeded && Kernel32.GetCurrentActCtx(out IntPtr current).IsTrue())
+            {
+                return current == s_hActCtx;
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        ///  Activate() does nothing if a theming context is already active on the current thread, which is good
+        ///  for perf reasons. However, in some cases, like in the Timer callback, we need to put another context
+        ///  on the stack even if one is already present. In such cases, this method helps - you get to manage
+        ///  the cookie yourself though.
+        /// </summary>
+        public static IntPtr Activate()
```

Comment:
What do you think about this? (I cannot add a multiline suggestion)
```suggestion
        public static IntPtr Activate()
            {
                IntPtr userCookie = IntPtr.Zero;
                if (IsContextCreatedButNotActivated())
                {
                    userCookie = ActivateActCtx(hActCtx, out userCookie) ? userCookie : IntPtr.Zero;
                }
                return userCookie;
            }

            private static bool IsContextCreatedButNotActivated()
            {
                return Application.UseVisualStyles && contextCreationSucceeded && OSFeature.Feature.IsPresent(OSFeature.Themes) && !IsContextActive();
            }
```